ejb/WFLY-6143-Ability-to-add-a-server-side-configured-Interceptor-for-the-applications-without-changing-the-deployments.adoc

```diff
@@ -0,0 +1,57 @@
+= Ability to add a server-side configured Interceptor for the applications without changing the deployments
+:author:            Tomasz Adamski
+:email:             tomasz.adamski@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+The goal of this issue is to implement server side EJB interceptors that can be configured system-wide through ejb3 subsystem configuration.
+
+=== Issue
+* https://issues.jboss.org/browse/WFLY-6143
+
+=== Dev Contacts
+
+* mailto:tadamski@redhat.com
+
+=== QE Contacts
+
+=== Testing By
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+EJB
+
+== Requirements
+* Interceptors are implemented and packed inside the custom modules
+* Interceptors should be configured in the subsystem configuration and not in deployments
+* There should be a number of binding points in which the interceptors should be executed (f.e. after security)
+
+== Solution draft
+
+EJB3 subsystem is extended with <server-interceptors> tags, which specifies modules that include the interceptors implementation, interceptors that are going to be plugged and a binding point for each of them.
+Sample XML fragment:
+
+<server-interceptors>
+	<modules>
+		<module>org.abccorp:tracing-interceptors:1.0</module>
+		<module>org.abccorp:timers-interceptors:1.1</module>
+	<modules>
+	<interceptors>
+		<interceptor class="org.abccorp.TracingInterceptor" binding="post-security"/>
```

Comment:
@brianstansberry I'm OK with changing the so that the module is repeated in each interceptor tag. The management model would be as follows: I will add another child, server-interceptor, to ejb3 subsytem resource and each interceptor will be represented by it's own child resource.

Regarding other questions, I'm planning the following implementation: For each binding point create an interceptor (== org.jboss.invocation.Interceptor) implementation that would be responsible for executing customer interceptors for this particular binding. In the context of that:
1. Binding points: There can be arbitrary number of binding points. otoh when the above implementation is ready adding another binding point should be simple. As a result of that, my initial idea was to start with one binding point (f.e. "default", at the beginning of invocation chain) and later add binding points based on customer demand. This is not an ideal solution, but it seems to me better than choosing arbitrary group of points by myself.
2. Restarting the services: I'm still reserching the implementation but my initial preference would be to make the changes to server interceptors induce reload-required. Changing interceptors in the running server would imo induce subystantial implementation complexity (f.e. modifing deployment module dependcies in the flight, modifing components view in the flight with all the possible race conditions etc.) so imo we would need strong customer demand for that.

Reply:
@tadamski any update on this? Should I consider your last commentary as final implementation draft?